server_development/topics/auth-spi.adoc

```diff
@@ -108,199 +111,547 @@ Let's walk through the steps from when a client first redirects to keycloak to a
-        if (!validated) {
-           Response challenge = context.form()
-                                 .setError("badSecret")
-                                 .createForm("secret-question.ftl");
-           context.failureChallenge(AuthenticationFlowError.INVALID_CREDENTIALS, challenge);
-           return;
-        }
-        setCookie(context);
-        context.success();
-    }
-    
-    protected boolean validateAnswer(AuthenticationFlowContext context) {
-        MultivaluedMap<String, String> formData = context.getHttpRequest().getDecodedFormParameters();
-        String secret = formData.getFirst("secret_answer");
-        UserCredentialModel input = new UserCredentialModel();
-        input.setType("SECRET_QUESTION");
-        input.setValue(secret);
-        return context.getSession().userCredentialManager().isValid(context.getRealm(), context.getUser(), input);
+@Override
+public void action(AuthenticationFlowContext context) {
+    boolean validated = validateAnswer(context);
+    if (!validated) {
+        Response challenge =  context.form()
+                .setError("badSecret")
+                .createForm("secret-question.ftl");
+        context.failureChallenge(AuthenticationFlowError.INVALID_CREDENTIALS, challenge);
+        return;
     }
-----            
+    setCookie(context);
+    context.success();
+}
+----
 
 If the answer is not valid, we rebuild the HTML Form with an additional error message.
 We then call AuthenticationFlowContext.failureChallenge() passing in the reason for the value and the JAX-RS response.
-failureChallenge() works the same as challenge(), but it also records the failure so it can be analyzed by any attack detection service. 
+failureChallenge() works the same as challenge(), but it also records the failure so it can be analyzed by any attack detection service.
+
+If validation is successful, then we set a cookie to remember that the secret question has been answered and we call AuthenticationFlowContext.success().
 
-If validation is successful, then we set a cookie to remember that the secret question has been answered and we call AuthenticationFlowContext.success(). 
+The validation itself gets the data that was received from the form, and calls the isValid method from the SecretQuestionCredentialProvider. You'll notice
+that there's a section of the code concerning getting the credential Id. This is because if {project_name} is configured to allow multiple types of alternative
+authenticators, or if the user could record multiple credentials of the SECRET_QUESTION type (for example if we allowed to choose from several questions,
+and we allowed the user to have answers for more than one of those questions), then {project_name} needs to know which credential is being user to log the user.
+In case there is more than one, {project_name} allows the user to choose during the login which credential he is using, and the information is transmitted by
```

Comment:
make it "more than one credential"
also change "he is using" to "is being used"